keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -871,32 +878,110 @@ not added to the pod for a long enough time (for example 2 minutes).
+but is doomed to be failed, as it is terminating (has `deletionTimestamp` set).
+In order to workaround this issue, Job controller, when pod failure policy is
+disabled, considers any terminating pod that is in a non-terminal phase as failed.
+Note that, it is important that when Job controller considers such pods as failed
+so that it removes their finalizers and thus allows the API server to complete
+their deletion.
+
+In order to ensure consistency in behavior when handling pod failures when
+pod failure policy is used or not, we need to make sure that all pods which
+are terminating (were previously considered as failed by the Job controller),
+are transitioned to the failed state eventually.
+
+Thus, we review scenarios in which a pod is considered by Job controller as
+failed, but may get stuck in a non-terminal.
+
+1. **Orphan pods** (solved in Alpha)
+
+This pod state is characterized by the following:
+- scheduled (the Pod's `.spec.nodeName` is set), but the node no longer exists
+- `Running` or `Pending` phase
+
+Note that, as PodGC sends DELETE request for such pod (DELETE request was sent
+prior to this KEP, but without setting the phase to `Failed`), it becomes also
+terminating (has `deletionTimestamp` set).
+
+Example scenario leading to the Pod's state:
+- pod scheduled to a node, might be `Running` or `Pending`
+- node deleted (see: [Disconnected node](#disconnected-node))
+
+2. **Pending, terminating and unscheduled** (solved in Alpha)
+
+This pod state is characterized by the following:
+- unscheduled (the Pod's `.spec.nodeName` is nil), so no Kubelet is assigned
+- `Pending` phase
+- the Pod's `deletionTimestamp` is set by a DELETE request (terminating)
+
+Example scenario leading to the Pod's state:
+- unschedulable pod (for example due to unsatisfiable requests)
+- DELETE request sent by a user or another k8s component
+
+Note that, the point about the Pod being terminating is important here. As
+long as the pod is not terminating it can be scheduled if resources are
+increased to satisfy its requests.
+
+3. **Pending, terminating, and scheduled** (planned for second Beta)
+
+This pod state is characterized by the following:
+- the Pod's `.spec.nodeName` is set, but the node no longer exists
+- `Pending` phase
+- the Pod's `deletionTimestamp` is set by a DELETE request (terminating)
+
+Example scenario leading to the Pod's state:
+- Pod scheduled to a node, but remains in `Pending` (e.g. due to an invalid
+image reference, invalid config map, init container failing, one of the containers
+failing to start)
+- DELETE request sent by a user or another k8s component
+
+Note that, the point about the Pod being terminating is important here. As long
+as the pod is not terminating, it could be a transient issue and Kubelet can
+make progress after retrying. Once such a Pod is also terminating, Kubelet is not
+going to attempt starting of the pod.
+
+##### Implementation progress and plan
+
```

Comment:
A couple items I think we should discuss in the KEP:

1. When the kubelet's sets the phase to failed for these pods, what should the message be communicated on pod status? i.e. phase=Failed, but what about Status and Message [1]? Ideally it would contain some info of why the pod was stuck in pending to help users debug.
2. We should ideally ensure that the kubelet reports the latest `ContainerStatuses` on the pod as well as part of this last update. If not, we have some bug in kubelet we need to fix.
3. I chatted about this change briefly with @smarterclayton. Clayton mentioned it is arguably a bug that this not happen today, i.e. pods are expected to go through their entire lifecycle and end at Succeeded or Failed. So it might be worth to call out in the KEP, that one of the reasons for this change is we want to ensure that all pods will end up in terminal state, even if they did not start (i.e. were stuck in pending).


[1] https://github.com/kubernetes/kubernetes/blob/49ddc510bc6b57fa430f779372b71144d328995c/staging/src/k8s.io/api/core/v1/types.go#L4023-L4027

Reply:
Thanks for the comment, I will update the KEP with a review of a couple of scenarios leading to the state. 
However, answering the comment here:
[1] we may need to end up with a generic messages such as {Reason: Evicted, Message: Evicted as pending and terminating}. The reason is that the Pending state would not necessarily make it stuck by itself. The reason for the DELETE request should be added in a pod condition by an external controller. For example, scheduler, will add `DisruptionTarget` condition with a message saying that it is preempting the pod. Still, the pod could get stuck in `Pending` phase, even without the DELETE, for example due to invalid image reference, but I think such scenarios deserve a dedicated KEP, because it might be hard to discriminate between cases where the image is misconfigured or there are transient connectivity issues.
[2] I'm intending to use the same scheme for evicting pods as currently used for example for pods which exceed the activeDeadlineSeconds timeout. This reports the `containerStatuses`.
[3] I agree with this, maybe it could be handled as a bugfix outside of the KEP (it is needed for the KEP, but maybe something else). Still, the KEP creates a nice communication channel to align on the approach of fixing it . Once done, we could consider cherry-picking it for 1.26.

Reply:
I think the reason should be `Deleted` (something deleted the pod). And the message could be "Deleted while Pending". But..... can we leave this discussion to code review? It sounds minor.

Reply:
@bobbypage updated the KEP with suggestions for [1] and [3] and provided a short review of example scenarios leading to the state. PTAL. As for [2] I think this could be an implementation thing.

Reply:
Thank you for all the updates in the KEP lgtm!